base/iddict.jl

```diff
@@ -83,18 +84,31 @@ function sizehint!(d::IdDict, newsz)
     rehash!(d, newsz)
 end
 
-function setindex!(d::IdDict{K,V}, @nospecialize(val), @nospecialize(key)) where {K, V}
+function ht_keyindex!(d::IdDict{K, V}, @nospecialize(key)) where {K, V}
     !isa(key, K) && throw(ArgumentError("$(limitrepr(key)) is not a valid key for type $K"))
-    if !(val isa V) # avoid a dynamic call
-        val = convert(V, val)::V
-    end
+    keyindex = ccall(:jl_eqtable_keyindex, Cssize_t, (Any, Any), d, key)
```

Comment:
why does this need to be implemented in C?

Reply:
I'm not sure how to call `jl_table_assign_bp` directly from Julia as it can change `d.ht`.

Reply:
There used to be ways, but not anymore with the generational GC. The usual way now to implement this is to switch the return object around as such:
```suggestion
    keyindex = Ref(UInt(0))
    d.ht = ccall(:jl_table_assign_bp, Any, (Any, Any, Ref{UInt}), d.ht, key, keyindex)
    keyindex = keyindex[]
```

Reply:
Can you change this around still also? It will avoid us from needing to reflect the iddict struct into C

Reply:
Just to note, I believe this would remove the ~20% speedup (see comments above) we get from not having to heap allocate for `setindex!`. Happy to make this change if that's fine.